File: FastMines_uwp/src/fmg/uwp/img/win2d/Logo.cs
using System.Linq;
using System.Collections.Generic;
using Windows.UI;
using Windows.UI.Xaml;
using Windows.Graphics.Display;
using Microsoft.Graphics.Canvas;
using Microsoft.Graphics.Canvas.UI.Xaml;
using Microsoft.Graphics.Canvas.Geometry;
using Fmg.Common;
using Fmg.Common.Geom;
using Fmg.Core.Img;
using Fmg.Uwp.Utils;
using Fmg.Uwp.Utils.Win2d;
using Fmg.Uwp.Mosaic.Win2d;

namespace Fmg.Uwp.Img.Win2d {

    /// <summary> Main logo image. Win2D implementation </summary>
    public static class Logo {

        /// <summary> Main logo image. Base view Win2D implementation </summary>
        /// <typeparam name="TImage">Win2D specific image: <see cref="CanvasBitmap"/> or <see cref="CanvasImageSource"/></typeparam>
        public abstract class Win2DView<TImage> : ImageView<TImage, LogoModel>
            where TImage : DependencyObject, ICanvasResourceCreator
        {

            protected readonly ICanvasResourceCreator _rc;

            protected Win2DView(ICanvasResourceCreator resourceCreator)
                : base(new LogoModel())
            {
                _rc = resourceCreator;
            }

            protected void Draw(CanvasDrawingSession ds, bool fillBk) {
                ICanvasResourceCreator rc = Image;
                LogoModel lm = Model;

                if (fillBk)
                    ds.Clear(lm.BackgroundColor.ToWinColor());

                IList<PointDouble> rays = lm.Rays;
                IList<PointDouble> inn = lm.Inn;
                IList<PointDouble> oct = lm.Oct;

                var center = new PointDouble(Size.Width / 2.0, Size.Height / 2.0);

                HSV[] hsvPalette = lm. Palette;
                Windows.UI.Color[] palette = hsvPalette
                    .Select(hsv => hsv.ToWinColor())
                    .ToArray();

                // paint owner rays
                for (var i = 0; i < 8; i++) {
                    using (var geom = rc.BuildLines(rays[i], oct[i], inn[i], oct[(i + 5) % 8])) {
                        if (!lm.UseGradient) {
                            ds.FillGeometry(geom, hsvPalette[i].ToColor().Darker().ToWinColor());
                        } else {
                            // emulate triangle gradient (see BmpLogo.cpp C++ source code)
                            // over linear gragients

                            using (var br = rc.CreateGradientPaintBrush(rays[i], palette[(i+1)%8], inn[i], palette[(i+6)%8])) {
                                ds.FillGeometry(geom, br);
                            }

                            var p1 = oct[i];
                            var p2 = oct[(i + 5) % 8];
                            var p = new PointDouble((p1.X + p2.X) / 2, (p1.Y + p2.Y) / 2); // середина линии oct[i]-oct[(i+5)%8]. По факту - пересечение линий rays[i]-inn[i] и oct[i]-oct[(i+5)%8]

                            Windows.UI.Color clr;// = new Color(255,255,255,0); //  Fmg.Common.Color.Transparent.ToWinColor();
                            if (true) {
                                HSV c1 = hsvPalette[(i + 1) % 8];
                                HSV c2 = hsvPalette[(i + 6) % 8];
                                double diff = c1.h - c2.h;
                                HSV cP = new HSV(c1.ToColor());
                                cP.h += diff / 2; // цвет в точке p (пересечений линий...)
                                cP.a = 0;
                                clr = cP.ToColor().ToWinColor();
                            }

                            using (var br = rc.CreateGradientPaintBrush(oct[i], palette[(i + 3) % 8], p, clr)) {
                                using (var geom2 = rc.BuildLines(rays[i], oct[i], inn[i])) {
                                    ds.FillGeometry(geom2, br);
                                }
                            }
                            using (var br = rc.CreateGradientPaintBrush(oct[(i + 5) % 8], palette[(i + 0) % 8], p, clr)) {
                                using (var geom2 = rc.BuildLines(rays[i], oct[(i + 5) % 8], inn[i])) {
                                    ds.FillGeometry(geom2, br);
                                }
                            }
                        }
                    }
                }

                // paint star perimeter
                var zoomAverage = (lm.ZoomX + lm.ZoomY) / 2;
                var penWidth = Model.BorderWidth*zoomAverage;
                if (penWidth > 0.1)
                    using (var css = new CanvasStrokeStyle {
                        StartCap = CanvasCapStyle.Round,
                        EndCap = CanvasCapStyle.Round
                    }) {
                        for (var i = 0; i < 8; i++) {
                            var p1 = rays[(i + 7) % 8];
                            var p2 = rays[i];
                            ds.DrawLine(p1.ToVector2(), p2.ToVector2(), palette[i], (float)penWidth, css);
                        }
                    }

                // paint inner gradient triangles
                for (var i = 0; i < 8; i++) {
                    using (var geom = rc.BuildLines(inn[(i + 0) % 8], inn[(i + 3) % 8], center)) {
                        if (lm.UseGradient) {
                            var p1 = inn[(i + 0) % 8];
                            var p2 = inn[(i + 3) % 8];
                            var p = new PointDouble((p1.X + p2.X) / 2, (p1.Y + p2.Y) / 2); // center line of p1-p2
                            using (var br = rc.CreateGradientPaintBrush(p, palette[(i + 6) % 8], center, ((i & 1) == 1) ? Colors.Black : Colors.White)) {
                                ds.FillGeometry(geom, br);
                            }
                        } else {
                            ds.FillGeometry(geom, ((i & 1) == 1)
                                ? hsvPalette[(i + 6) % 8].ToColor().Brighter().ToWinColor()
                                : hsvPalette[(i + 6) % 8].ToColor().Darker().ToWinColor());
                        }
                    }
                }
            }

            protected override void Disposing() {
                base.Disposing();
                Model.Dispose();
            }

        }

        /////////////////////////////////////////////////////////////////////////////////////////////////////
        //    custom implementations
        /////////////////////////////////////////////////////////////////////////////////////////////////////

        /// <summary> Logo image view implementation over <see cref="CanvasBitmap"/> </summary>
        public class CanvasBmpView : Win2DView<CanvasBitmap> {

            public CanvasBmpView(ICanvasResourceCreator resourceCreator)
                : base(resourceCreator)
            { }

            protected override CanvasBitmap CreateImage() {
                var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
                var s = Model.Size;
                return new CanvasRenderTarget(_rc, (float)s.Width, (float)s.Height, dpi);
            }

            protected override void DrawBody() {
                using (var ds = ((CanvasRenderTarget)Image).CreateDrawingSession()) {
                    Draw(ds, true);
                }
            }

        }

        /// <summary> Logo image view implementation over <see cref="CanvasImageSource"/> (XAML <see cref="Windows.UI.Xaml.Media.ImageSource"/> compatible) </summary>
        public class CanvasImgSrcView : Win2DView<CanvasImageSource> {

            public CanvasImgSrcView(ICanvasResourceCreator resourceCreator /* = CanvasDevice.GetSharedDevice() */)
                : base(resourceCreator)
            { }

            protected override CanvasImageSource CreateImage() {
                var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
                var s = Model.Size;
                return new CanvasImageSource(_rc, (float)s.Width, (float)s.Height, dpi);
            }

            protected override void DrawBody() {
                using (var ds = Image.CreateDrawingSession(Model.BackgroundColor.ToWinColor())) {
                    Draw(ds, false);
                }
            }

        }

        /// <summary> Logo image controller implementation for <see cref="Logo.CanvasBmpView"/> </summary>
        public class CanvasBmpController : LogoController<CanvasBitmap, Logo.CanvasBmpView> {

            public CanvasBmpController(ICanvasResourceCreator resourceCreator)
                : base(new Logo.CanvasBmpView(resourceCreator))
            { }

            protected override void Disposing() {
                base.Disposing();
                View.Dispose();
            }

        }

        /// <summary> Logo image controller implementation for <see cref="Logo.CanvasImgSrcView"/> </summary>
        public class CanvasImgSrcController : LogoController<CanvasImageSource, Logo.CanvasImgSrcView> {

            public CanvasImgSrcController(ICanvasResourceCreator resourceCreator)
                : base(new Logo.CanvasImgSrcView(resourceCreator))
            { }

            protected override void Disposing() {
                base.Disposing();
                View.Dispose();
            }

        }

    }

}

File: FastMines_swing/src/main/java/fmg/swing/img/Logo.java
package fmg.swing.img;

import java.awt.*;
import java.awt.geom.Point2D;
import java.util.Arrays;
import java.util.List;

import fmg.common.HSV;
import fmg.common.geom.PointDouble;
import fmg.core.img.LogoController2;
import fmg.core.img.LogoModel2;
import fmg.swing.utils.Cast;

/** Main logo image */
public final class Logo2 {
    private Logo2() {}

    private static final Color TRANSPARENT = new Color(0xFF, 0xFF, 0xFF, 0);

    private static void draw(Graphics2D g, LogoModel2 lm) {
        var size = lm.getSize();

        { // fill background
            g.setComposite(AlphaComposite.Src);
            g.setColor(TRANSPARENT);
            g.fillRect(0, 0, (int)size.width, (int)size.height);
        }

        g.setComposite(AlphaComposite.SrcOver);
        List<PointDouble> rays0 = lm.getRays();
        List<PointDouble> inn0  = lm.getInn();
        List<PointDouble> oct0  = lm.getOct();

        Point2D.Double[] rays = rays0.stream().map(Cast::toPoint).toArray(s -> new Point2D.Double[s]);
        Point2D.Double[] inn  = inn0 .stream().map(Cast::toPoint).toArray(s -> new Point2D.Double[s]);
        Point2D.Double[] oct  = oct0 .stream().map(Cast::toPoint).toArray(s -> new Point2D.Double[s]);
        Point2D.Double center = new Point2D.Double(size.width/2.0, size.height/2.0);

        HSV[] hsvPalette = lm.getPalette();
        Color[] palette = Arrays.stream(hsvPalette)
            .map(hsv -> Cast.toColor(hsv.toColor()))
            .toArray(s -> new Color[s]);

        // paint owner gradient rays
        for (int i=0; i<8; i++) {
            if (!lm.isUseGradient()) {
                g.setColor(Cast.toColor(hsvPalette[i].toColor().darker()));
                fillPolygon(g, rays[i], oct[i], inn[i], oct[(i+5)%8]);
            } else {
                // emulate triangle gradient (see BmpLogo.cpp C++ source code)
                // over linear gragients

                g.setPaint(new GradientPaint(rays[i], palette[(i+1)%8], inn[i], palette[(i+6)%8]));
                fillPolygon(g, rays[i], oct[i], inn[i], oct[(i+5)%8]);

                Point2D.Double p1 = oct[i];
                Point2D.Double p2 = oct[(i+5)%8];
                Point2D.Double p = new Point2D.Double((p1.getX()+p2.getX())/2, (p1.getY()+p2.getY())/2); // середина линии oct[i]-oct[(i+5)%8]. По факту - пересечение линий rays[i]-inn[i] и oct[i]-oct[(i+5)%8]

                Color clr;// = new Color(255,255,255,0); //  Cast.toColor(fmg.common.Color.Transparent);
                if (true) {
                    HSV c1 = hsvPalette[(i+1)%8];
                    HSV c2 = hsvPalette[(i+6)%8];
                    double diff = c1.h - c2.h;
                    HSV cP = new HSV(c1.toColor());
                    cP.h += diff/2; // цвет в точке p (пересечений линий...)
                    cP.a = 0;
                    clr = Cast.toColor(cP.toColor());
                }

                g.setPaint(new GradientPaint(oct[i], palette[(i+3)%8], p, clr));
                fillPolygon(g, rays[i], oct[i], inn[i]);

                g.setPaint(new GradientPaint(oct[(i+5)%8], palette[(i+0)%8], p, clr));
                fillPolygon(g, rays[i], oct[(i+5)%8], inn[i]);
            }
        }

        // paint star perimeter
        double zoomAverage = (lm.getZoomX() + lm.getZoomY())/2;
        final double penWidth = lm.getBorderWidth() * zoomAverage;
        if (penWidth > 0.1) {
            g.setStroke(new BasicStroke((float)penWidth, BasicStroke.CAP_ROUND, BasicStroke.JOIN_BEVEL));
            for (int i=0; i<8; i++) {
                Point2D.Double p1 = rays[(i + 7)%8];
                Point2D.Double p2 = rays[i];
                g.setColor(palette[i].darker());
                g.drawLine((int)p1.x, (int)p1.y, (int)p2.x, (int)p2.y);
            }
        }

        // paint inner gradient triangles
        for (int i=0; i<8; i++) {
            if (lm.isUseGradient()) {
                Point2D.Double p1 = inn[(i+0)%8];
                Point2D.Double p2 = inn[(i+3)%8];
                Point2D.Double p = new Point2D.Double((p1.getX()+p2.getX())/2, (p1.getY()+p2.getY())/2); // center line of p1-p2
                g.setPaint(new GradientPaint(
                    p, palette[(i+6)%8],
                    center, ((i & 1) == 1) ? Color.BLACK : Color.WHITE));
            } else
                g.setColor(((i & 1) == 1)
                    ? Cast.toColor(hsvPalette[(i + 6)%8].toColor().brighter())
                    : Cast.toColor(hsvPalette[(i + 6)%8].toColor().darker()));
            fillPolygon(g, inn[(i + 0)%8], inn[(i + 3)%8], center);
        }
    }

    private static void fillPolygon(Graphics2D g, Point2D.Double... p) {
        g.fillPolygon(
            Arrays.stream(p).mapToInt(s -> (int)s.x).toArray(),
            Arrays.stream(p).mapToInt(s -> (int)s.y).toArray(),
            p.length);
    }


    /** Logo image controller implementation for {@link javax.swing.Icon} */
    public static class LogoSwingIconController extends LogoController2<javax.swing.Icon, SwingIconView<LogoModel2>> {

        public LogoSwingIconController() {
            var model = new LogoModel2();
            var view = new SwingIconView<>(model, g -> Logo2.draw(g, model));
            init(model, view);
        }

    }

    /** Logo image controller implementation for {@link java.awt.Image} */
    public static class LogoAwtImageController extends LogoController2<java.awt.Image, AwtImageView<LogoModel2>> {

        public LogoAwtImageController() {
            var model = new LogoModel2();
            var view = new AwtImageView<>(model, g -> Logo2.draw(g, model));
            init(model, view);
        }

    }

}
